add opentelemetry

tobert opened the pull request:
## Description

Implements OpenTelemetry tracing in OSIE. This includes some code, otel-cli spans in shell scripts, and W3C traceparent propagation in http headers where we can.

## Why is this needed

When everything is in place, we will get detailed traces of operations happening in OSIE, making it observable.

## How are existing users impacted? What migration steps/scripts do we need?

Understanding what is happening in complex provisioning systems is difficult. We mostly get the job done with logs and metrics these days. This is an attempt to add OpenTelemetry events so that users of OSIE and other Tinkerbell components can emit traces to their favorite OpenTelemetry-compatible provider (or translate through opentelemetry-collector).

## Checklist:

My TODO:

- [x] add otel-cli to the image build process
- [x] pass TRACEPARENT environment variables through docker calls
- [x] add W3C traceparent headers to curl calls in scripts
- [x] add opentelemetry-python to osie-runner (or at least propagate traceparent headers)
- [x] figure out the best way to configure otel-cli at runtime (propagate OTEL_EXPORTER_OTLP_*)

Related work:

- [ ] add TRACEPARENT envvar support to `tink` (WIP)
- [ ] teach boots http to accept traceparent as a query param (TBD)
- [x] merge boots PR to add tp in kernel command line
- [x] finish up otel-cli status support (https://github.com/equinix-labs/otel-cli/pull/27) (handy for reporting failures)
- [x] make `otel-cli span background --tp-carrier` write out a file that can be sourced
- [x] add `otel-cli span background --wait`

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

Commit:
add otel-cli to the Dockerfile

A slight departure from what's there before, this uses a docker COPY
--from= to grab the otel-cli container and grab the otel-cli binary from
there.

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
add opentelemetry dependencies to osie-runner

seems pretty reasonable, most of the lines are otel referencing itself,
pleasingly few other dependencies

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
[wip] first pass at adding otel

this compiles and starts and I can see the otel instrumentation in the
stack trace I get when phone-home fails

as it says in the TODO there is more to do but I am done for today and
this is committable

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
propagate traceparent globally

Figured this out properly, the thing I needed was a context.attach(ctx).

Of course, this kind of stuff isn't very well documented.

Validated with `otel-cli server tui` and:

```sh
otel-cli span --tp-print |tee  x && . x && export TRACEPARENT

docker run \
	--network=host \
	-e TRACEPARENT=$TRACEPARENT \
	-e OTEL_EXPORTER_OTLP_INSECURE=true \
	-e OTEL_EXPORTER_OTLP_ENDPOINT=localhost:4317 \
	-e STATEDIR_HOST=/tmp/tmp.6QrN4tjUSh \
	-v $(pwd)/run.py:/run.py \
	-ti osie-runner
```

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
Merge branch 'master' into opentelemetry-take-1

Commit:
add some otel hooks

some otel-cli

some curl headers

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
use long arg names

Signed-off-by: Amy Tobey <atobey@equinix.com>

splaspood commented on osie-runner/run.py:
```diff
@@ -102,6 +138,9 @@ def sanitize_cacher_data(j):
     cmdline_content = cmdline.read()
     tinkerbell = parse.urlparse(util.value_from_kopt(cmdline_content, "tinkerbell"))
     facility = util.value_from_kopt(cmdline_content, "facility")
+    kernel_tp = util.value_from_kopt(cmdline_content, "traceparent") # opentelemetry
```
Wondering if maybe we want to shorten this kopt.  I keep going back and forth on it (it's not really a big deal either way) and maybe this comment is better made on the corresponding boots PR.

tobert replied:
I seriously considered that, but given how new otel is to most folks, let alone those of us in bare metal land, I figured it was better to keep the long name for clarity when people ask "wtf is this?"

Commit:
bump to otel-cli v0.0.15

fixed a ridiculous bug in otel-cli where it didn't actually read
environment variables, sigh

also added `otel-cli span background --wait` which will make the usage
in docker/scripts/functions.sh work as expected

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
bump to otel-cli v0.0.16

I forgot I didn't have --wait merged into v0.0.15 so I went off and got
that polished off in these PRs:

https://github.com/equinix-labs/otel-cli/pull/73
https://github.com/equinix-labs/otel-cli/pull/74

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
add otel-cli span to functions.sh/init()

Most/all of the implementation scripts start out by calling
`source functions.sh && init` so it's a natural place to put a hook into
otel-cli.

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
only try to source files that exist

It should be incredibly rare / impossible for otel-cli to not write out
the carrier file via --tp-carrier, but given the sensitivity of this
funciton, it seems best to be defensive and fail silently when that
happens.

Signed-off-by: Amy Tobey <atobey@equinix.com>

tobert committed:
Merge branch 'tinkerbell:main' into opentelemetry-take-1

Commit:
generate a traceparent if none was found, return it

I was plumbing traceparent down to the docker commands and started
adding `if traceparent is None` checks and realized that's not great to
copy around and it would be better if traceparent is always populated
and safe to use, so this does that.

When no traceparent is in kernel command line or in the environment, it
will start a fresh unparented span and return it as the traceparent.

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
plumb otel traceparent through to docker run

Adds traceparent plumbing to Handler, adding it as a member and
appending it to the environments passed to `run_osie` at each callsite.
I did it this way because for some reason `run_osie` is a static method
and does not have a `self` so this gets everything we want well enough
without changing that.

Commit:
add OTLP envvars and move envvar settings to a helper

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
shellcheck

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
lint

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
fix tests

Since the otel stuff is non-functional I used the handler's
inject_otel_envvars method to sort things out. This isn't ideal from a
mock testing perspective but does get the tests passing with minimal
change.

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
lint

this is ridiculous

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
add traceparent headers to failures phoning home

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
remove accidental comma

shellcheck complained about inlining but not this, thank you shellcheck

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
whitespace for linter

this is very fun!

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
bump to otel-cli v0.0.17

to remove log line that was somehow breaking the cpr thing that tees
script output

this shouldn't be happening since log.Printf in Go should go to stderr
but whatever, it's gone now

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
use 30m instead of 1800

This failed on previous version of otel-cli because of a bug in otel-cli
that is fixed in v0.0.17 that refused bare numbers of seconds. That's
fixed and also now this specifies minutes which is a bit easier to read.

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
wrap $TRACEPARENT in ${TRACEPARENT:-} for set -u

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
bump to otel-cli v0.0.18

Fixed the thing where it's printing the traceparent to stdout which was
very bad.

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
Merge branch 'main' into opentelemetry-take-1

Commit:
Merge branch 'main' into opentelemetry-take-1

Commit:
also send version as an attribute

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
add more traceparent passing to curls and dockers

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
Merge branch 'master' into opentelemetry-take-1

Signed-off-by: Amy Tobey <atobey@equinix.com>

Commit:
fix bad merge

Whatever happened with the previous change it interacted really poorly
with my PR and somehow VS code thought the deleted code here was OK to
commit!? Who knows, this is my 3rd attempt and this is fine
:thisisfinedog:

Signed-off-by: Amy Tobey <atobey@equinix.com>

jacobweinstock committed:
Merge branch 'main' into opentelemetry-take-1

jacobweinstock commented:
Hey @ScottGarman , @splaspood would either of you be available, by chance, to check this out?

ScottGarman commented:
I'm :+1: on the overall gist of this, though I do wonder if it would be simpler to start with a specific component (like osie-runner) to instrument first, then roll it out more widely. Other than that, just request typical commit cleanup/squashes and a rebase to the main branch. I can then assist with the process of deploying an osie test build so we can run this in production at a limited scale.

jacobweinstock committed:
Merge branch 'main' into opentelemetry-take-1

tobert commented:
@ScottGarman I have been thinking about splitting this PR and might just do that. Getting the otel-cli bits to work right will need more work to pass around `OTEL_EXPORTER_OTLP_ENDPOINT` or some other configuration pointing at a server, so I think we can defer the otel-cli bits a while and see how things look after we light all this up.

tobert commented:
I know y'all don't like to leave these hanging. Gonna close this and will open new PRs soon with the same code split up.
